docs: migration from v2 guide (#8390)
vitejs/vite: docs/guide/migration.md

# Migration from v2

## Node Support

> Vite no longer supports Node v12, which reached its EOL. Node 14.4+ is now required.

Reviewer: ```suggestion
Vite no longer supports Node v12, which reached its EOL. Node 14.6+ is now required.
```

https://github.com/vitejs/vite/blob/842f995ca69600c4c06c46d202fe713b80373418/packages/vite/package.json#L37

## Modern Browser Baseline change

The production bundle assumes support for modern JavaScript. By default, Vite targets browsers which support the [native ES Modules](https://caniuse.com/es6-module) and [native ESM dynamic import](https://caniuse.com/es6-module-dynamic-import) and [`import.meta`](https://caniuse.com/mdn-javascript_statements_import_meta):

- Chrome >=87
- Firefox >=78
- Safari >=13
- Edge >=88

A small fraction of users will now require using [@vitejs/plugin-legacy](https://github.com/vitejs/vite/tree/main/packages/plugin-legacy), which will automatically generate legacy chunks and corresponding ES language feature polyfills.

> ## Config Options Change

Reviewer: ```suggestion
## Config Options Changes
```

or

```suggestion
## Config Options changed
```

Reviewer: ... all titles should stick to one convention
A third option would be to use `Title Case So All Words Start With Capital Latter`

- The following options that were already deprecated in v2 have been removed:

  - `optimizeDeps.keepNames` (switch to [`optimizeDeps.esbuildOptions.keepNames`](../config/dep-optimization-options.md#optimizedepsesbuildoptions))
  - `build.base` (switch to [`base`](../config/shared-options.md#base))
  - `alias` (switch to [`resolve.alias`](../config/shared-options.md#resolvealias))
  - `dedupe` (switch to [`resolve.dedupe`](../config/shared-options.md#resolvededupe))
  - `polyfillDynamicImport` (use [`@vitejs/plugin-legacy](https://github.com/vitejs/vite/tree/main/packages/plugin-legacy) for browsers without dynamic import support)

> ## Dev server Changes

Reviewer: ```suggestion
## Dev Server changes
```

Reviewer 2: Shouldn't this be `Dev Server Changes` for consistency with the rest of the document?

```suggestion
## Dev Server Changes
```

> Vite optimizes dependencies with esbuild to both convert CJS only deps to ESM and to reduce the number of modules the browser needs to request. In v3 the default strategy to discover and batch dependencies has changed. Vite no longer pre-scan user code with esbuild to get an initial list of dependencies on cold start. Instead, it delays the first dependency optimization run until every imported user module on load is processed. To get back the v2 strategy you can use [`optimizeDeps.scan`](../config/dep-optimization-options.md#optimizedepsscan).

Reviewer 3: ```suggestion
Vite optimizes dependencies with esbuild to both convert CJS only deps to ESM and to reduce the number of modules the browser needs to request. In v3 the default strategy to discover and batch dependencies has changed. Vite no longer pre-scan user code with esbuild to get an initial list of dependencies on cold start. Instead, it delays the first dependency optimization run until every imported user module on load is processed. To get back the v2 strategy you can use [`optimizeDeps.devScan`](../config/dep-optimization-options.md#optimizedepsdevscan).
```

> Vite optimizes dependencies with esbuild to both convert CJS only deps to ESM and to reduce the number of modules the browser needs to request. In v3 the default strategy to discover and batch dependencies has changed. Vite no longer pre-scan user code with esbuild to get an initial list of dependencies on cold start. Instead, it delays the first dependency optimization run until every imported user module on load is processed. To get back the v2 strategy you can use [`optimizeDeps.devScan`](../config/dep-optimization-options.md#optimizedepsdevscan).

Reviewer: Maybe add some line breaks, to make it more readable?

Reviewer 3: done

> Vite optimizes dependencies with esbuild to both convert CJS only deps to ESM and to reduce the number of modules the browser needs to request. In v3 the default strategy to discover and batch dependencies has changed. Vite no longer pre-scan user code with esbuild to get an initial list of dependencies on cold start. Instead, it delays the first dependency optimization run until every imported user module on load is processed. To get back the v2 strategy you can use [`optimizeDeps.devScan`](../config/dep-optimization-options.md#optimizedepsdevscan).

Reviewer 2: ```suggestion
Vite optimizes dependencies with esbuild to both convert CJS-only deps to ESM and to reduce the number of modules the browser needs to request. In v3, the default strategy to discover and batch dependencies has changed. Vite no longer pre-scans user code with esbuild to get an initial list of dependencies on cold start. Instead, it delays the first dependency optimization run until every imported user module on load is processed. To get back the v2 strategy, you can use [`optimizeDeps.devScan`](../config/dep-optimization-options.md#optimizedepsdevscan).
```

> ## Build Changes

Reviewer: ```suggestion
## Build changes
```

Reviewer 2: @Shinigami92 Hmm. Why did you suggest title case in other places but the opposite here?

Reviewer: While scrolling through the PR I find out that the headlines doesn't use the same casing / style 
At the end one style should be chosen

> In v3, Vite uses esbuild to optimize dependencies by default. Doing so, it removes one of the most significant differences between dev and prod present in v2. Because esbuild converts CJS only dependencies to ESM, [`@rollupjs/plugin-commonjs`] is no longer used. If you need to get back to the v2 strategy, you can use [`optimizeDeps.disabled: 'build'`](../config/dep-optimization-options.md#optimizedepsdisabled).

Reviewer 2: ```suggestion
In v3, Vite uses esbuild to optimize dependencies by default. Doing so, it removes one of the most significant differences between dev and prod present in v2. Because esbuild converts CJS-only dependencies to ESM, [`@rollupjs/plugin-commonjs`] is no longer used. If you need to get back to the v2 strategy, you can use [`optimizeDeps.disabled: 'build'`](../config/dep-optimization-options.md#optimizedepsdisabled).
```

> ## SSR Changes

Reviewer: ...

> Vite v3 uses ESM for the SSR build by default. When using ESM, the [ssr externalization heuristics](https://vitejs.dev/guide/ssr.html#ssr-externals) are no longer needed. By default, all dependencies are externalized. You can use [`ssr.noExternal`](../config/ssr-options.md#ssrnoexternal) to control what dependencies to include in the SSR bundle. You can use `ssr.format: 'cjs'` to generate a CJS bundle, in which cases the same externalization strategy of Vite v2 is used.

Reviewer 2: ```suggestion
Vite v3 uses ESM for the SSR build by default. When using ESM, the [SSR externalization heuristics](https://vitejs.dev/guide/ssr.html#ssr-externals) are no longer needed. By default, all dependencies are externalized. You can use [`ssr.noExternal`](../config/ssr-options.md#ssrnoexternal) to control what dependencies to include in the SSR bundle. You can use `ssr.format: 'cjs'` to generate a CJS bundle, in which cases the same externalization strategy of Vite v2 is used.
```

> ## General Changes

Reviewer: ...

- [Raw `import.meta.glob`](features.md#glob-import-as) switched from `{ assert: { type: 'raw' }}` to `{ as: 'raw' }`

> - JS Extensions in SSR and lib mode now use a valid js extension (`js`, `mjs`, or `cjs`) for output js entries and chunks based on their format and the package type.

Reviewer 2: ```suggestion
- JS file extensions in SSR and lib mode now use a valid extension (`js`, `mjs`, or `cjs`) for output JS entries and chunks based on their format and the package type.
```
